gax/src/main/java/com/google/api/gax/tracing/TracedServerStreamingCallable.java

```diff
@@ -0,0 +1,138 @@
+ */
+package com.google.api.gax.tracing;
+
+import com.google.api.core.BetaApi;
+import com.google.api.core.InternalApi;
+import com.google.api.gax.rpc.ApiCallContext;
+import com.google.api.gax.rpc.ResponseObserver;
+import com.google.api.gax.rpc.ServerStreamingCallable;
+import com.google.api.gax.rpc.StreamController;
+import com.google.common.base.Preconditions;
+import javax.annotation.Nonnull;
+
+/**
+ * A wrapper callable that will wrap a callable chain in a trace.
+ *
+ * <p>This class is meant to be an internal implementation google-cloud-java clients only.
+ */
+@BetaApi("The surface for tracing is not stable and might change in the future")
+@InternalApi("For internal use by google-cloud-java clients only")
+public class TracedServerStreamingCallable<RequestT, ResponseT>
+    extends ServerStreamingCallable<RequestT, ResponseT> {
+
+  @Nonnull private final ApiTracerFactory tracerFactory;
+  @Nonnull private final SpanName spanName;
+  @Nonnull private final ServerStreamingCallable<RequestT, ResponseT> innerCallable;
+
+  public TracedServerStreamingCallable(
+      @Nonnull ServerStreamingCallable<RequestT, ResponseT> innerCallable,
+      @Nonnull ApiTracerFactory tracerFactory,
+      @Nonnull SpanName spanName) {
+    this.tracerFactory = Preconditions.checkNotNull(tracerFactory, "tracerFactory can't be null");
+    this.spanName = Preconditions.checkNotNull(spanName, "spanName can't be null");
+    this.innerCallable = Preconditions.checkNotNull(innerCallable, "innerCallable can't be null");
+  }
+
+  @Override
+  public void call(
+      RequestT request, ResponseObserver<ResponseT> responseObserver, ApiCallContext context) {
+
+    ApiTracer tracer = tracerFactory.newTracer(spanName);
+    TracedResponseObserver<ResponseT> tracedObserver =
+        new TracedResponseObserver<>(tracer, responseObserver);
+
+    context = context.withTracer(tracer);
+
+    try {
+      innerCallable.call(request, tracedObserver, context);
+    } catch (RuntimeException e) {
+      tracedObserver.onError(e);
+      throw e;
+    }
+  }
+
+  /** A {@link ResponseObserver} to mark a started operation trace as finished. */
+  private static class TracedResponseObserver<ResponseT> implements ResponseObserver<ResponseT> {
+    private final ApiTracer tracer;
+    private final ResponseObserver<ResponseT> innerObserver;
+    private volatile boolean wasCancelled;
+
+    private TracedResponseObserver(
+        @Nonnull ApiTracer tracer, @Nonnull ResponseObserver<ResponseT> innerObserver) {
+      this.tracer = Preconditions.checkNotNull(tracer, "tracer can't be null");
+      this.innerObserver = Preconditions.checkNotNull(innerObserver, "innerObserver can't be null");
+    }
+
+    @Override
+    public void onStart(final StreamController controller) {
+      innerObserver.onStart(
+          new StreamController() {
+            @Override
+            public void cancel() {
+              wasCancelled = true;
```

Comment:
This looks like the only difference between this internal static class and the bidistream version https://github.com/googleapis/gax-java/pull/636/files#diff-08e9aa9f0dcb325e4bd2a1770d907a77R98. In general it seems like a lot of duplication (sometimes completely identical code, like this example) between the internal static classes of the `Traced*Callable`. 

Please strongly consider. Moving the internal static classes to a package private classes under `tracing` package and reuse them in all of the `Traced*` callables.

Reply:
done